saleor/graphql/product/schema.py

```diff
@@ -336,8 +350,31 @@ def resolve_product(self, info, id=None, slug=None, channel=None, **_kwargs):
             )
         return ChannelContext(node=product, channel_slug=channel) if product else None
 
+    @staticmethod
     @traced_resolver
-    def resolve_products(self, info, channel=None, **kwargs):
+    def resolve_products(_root, info: graphene.ResolveInfo, channel=None, **kwargs):
+        # sort by RANK can be used only with search filter
+        if "sort_by" in kwargs and ProductOrderField.RANK == kwargs["sort_by"].get(
+            "field"
+        ):
+            if (
+                "filter" not in kwargs
+                or kwargs["filter"].get("search") is None
+                or not kwargs["filter"]["search"].strip()
+            ):
+                raise GraphQLError("Sorting by Rank is available only with searching.")
+        if (
+            "filter" in kwargs
+            and kwargs["filter"].get("search") is not None
+            and kwargs["filter"]["search"].strip()
+        ):
+            # default to sorting by rank unless specified otherwise
+            if "sort_by" not in kwargs:
+                product_type = info.schema.get_type("ProductOrder")
+                kwargs["sort_by"] = product_type.create_container(
+                    {"direction": "-", "field": ["search_rank", "id"]}
+                )
+
```

Comment:
Maybe we could wrap it with some function to make the resolver a bit easier to read?

Reply:
Extracted two helper functions for the `if` logic.